src/Analysis/Codelyzer.Analysis.CSharp/Handlers/ClassDeclarationHandler.cs

```diff
@@ -1,42 +1,78 @@
 using Codelyzer.Analysis.Common;
 using Codelyzer.Analysis.Model;
-using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis;
 using Microsoft.CodeAnalysis.CSharp.Syntax;
+using System.Collections.Generic;
 using System.Linq;
 
 namespace Codelyzer.Analysis.CSharp.Handlers
 {
     public class ClassDeclarationHandler : UstNodeHandler
     {
+        // Key: class full identifier, value: ClassDeclaration.
+        private Dictionary<string, ClassDeclaration> _classDeclarationCache = new();
         private ClassDeclaration ClassDeclaration { get => (ClassDeclaration)UstNode; }
 
         public ClassDeclarationHandler(CodeContext context,
             ClassDeclarationSyntax syntaxNode)
             : base(context, syntaxNode, new ClassDeclaration())
         {
             var classSymbol = SemanticHelper.GetDeclaredSymbol(syntaxNode, SemanticModel, OriginalSemanticModel);
-
-            ClassDeclaration.Identifier = syntaxNode.Identifier.ToString();
-            ClassDeclaration.Modifiers = syntaxNode.Modifiers.ToString();
-
             if (classSymbol != null)
             {
-                if(classSymbol.BaseType != null)
-                {
-                    ClassDeclaration.BaseType = classSymbol.BaseType.ToString();               
-                    ClassDeclaration.BaseTypeOriginalDefinition = GetBaseTypOriginalDefinition(classSymbol);
-                    ClassDeclaration.Reference.Namespace = GetNamespace(classSymbol);
-                    ClassDeclaration.Reference.Assembly = GetAssembly(classSymbol);
-                    ClassDeclaration.Reference.Version = GetAssemblyVersion(classSymbol);
-                    ClassDeclaration.Reference.AssemblySymbol = classSymbol.ContainingAssembly;
-                    ClassDeclaration.FullIdentifier = classSymbol.OriginalDefinition.ToString();
-                }
-                
-                if(classSymbol.Interfaces != null)
+                var fullIdentifier = GetFullIdentifier(classSymbol);
+                if (_classDeclarationCache.ContainsKey(fullIdentifier))
                 {
-                    ClassDeclaration.BaseList = classSymbol.Interfaces.Select(x => x.ToString())?.ToList();
+                    UstNode = _classDeclarationCache[fullIdentifier];
+                    return;
                 }
+                Set(ClassDeclaration, classSymbol);
+                _classDeclarationCache.Add(fullIdentifier, ClassDeclaration);
+            }
+        }
+
+        private void Set(ClassDeclaration classDeclaration, INamedTypeSymbol classSymbol)
+        {
+            classDeclaration.Identifier = classSymbol.Name;
```

Comment:
Yes, I believe both of them are the name of the class.

Reply:
How would this work if the semantic model is null?

Reply:
True, but note that previously if `classSymbol` was null, then `ClassDeclaration.Identifier` and `ClassDeclaration.Modifiers` still would have been set using values from `syntaxNode`

Reply:
That's true. Let me change it back to use the syntaxNode.

Reply:
I added the changes to use the identifier from the syntaxNode, but the tests failed. After I reverted it, the build passes now. Do you know why?

Reply:
Hi @marknfawaz @comsemer, I believe
1. The semantic model won't be null, because we get/create it from the compilation unit (the source file).
2. The `classSymbol` won't be null, because we get it from the `syntaxNode` which is a ClassDeclarationSyntax. Or, if it's the base class symbol, we check to make sure it's not null.

For your reference, https://docs.microsoft.com/en-us/dotnet/csharp/roslyn-sdk/get-started/semantic-analysis